pandas/tests/reshape/merge/test_merge.py

```diff
@@ -1437,6 +1437,13 @@ def test_merge_on_ints_floats(self, int_vals, float_vals, exp_vals):
         result = B.merge(A, left_on="Y", right_on="X")
         assert_frame_equal(result, expected[["Y", "X"]])
 
+    def test_merge_key_dtype_cast(self):
+        # GH 17044
+        df1 = pd.DataFrame({'key': [1., 2.], 'c1': [10, 20]})
+        df2 = pd.DataFrame({'key': [2], 'c2': [200]})
+        df = df1.merge(df2, on='key', how='left')
+        assert df['key'].dtype == 'float64'
```

Comment:
If you could `tm.assert_frame_equal` the resulting merged dataframe, that would be ideal (which checks the dtypes too.

```
result = df1.merge(df2, on='key', how='left')
expected = ....
tm.assert_frame_equal(result, expected)

```